---
title: "maize stress manuscript figures and tables"
output:
  pdf_document:
    citation_package: natbib
    keep_tex: false
    fig_caption: true
    latex_engine: lualatex
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
date: "`r format(Sys.time(), '%B %d, %Y')`"
fontsize: 11pt
geometry: "left=0.5in,right=0.5in,top=0.8in,bottom=0.5in"
# spacing: double
# biblio-style: apsr
header-includes:
- \PassOptionsToPackage{table}{xcolor}
- \usepackage{fontspec}
- \setmainfont[Scale=MatchUppercase]{TeX Gyre Heros}
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
---
\pagenumbering{gobble}

```{r, message=FALSE, echo=FALSE}
dirf = '/home/springer/zhoux379/projects/stress/data/95_figures/plots'
```

\hfill

![](plots/sf01.pdf){width=60%}

Figure S1. Experimental design used for generation of RNA-Seq data. (A) Temperature readings from a sensor that was with the plants throughout the experiment and shows the time of light / dark (gray shading indicates time without lights). (B-D) Three different experiments were conducted to assess the changes in gene expression in response to cold or heat stress in 14 day maize seedlings.  In experiment 1 (time course - TC, panel B) one replicate for three inbreds was collected at each of the indicated time-points.  In experiment 2 (inbred - hybrid - HY, panel C) three biological replicates were sampled from three maize inbreds and their F1 hybrids at time 0 and two time points during the stress.  In experiment 3 (panel D) a single biological replicate was sampled for a panel of 25 genotypes under control and cold conditions at two time points.

\pagebreak
![](plots/sf02.pdf){width=100%}

Figure S2. Hierarchical (A) and t-SNE (B) clustering of all samples from the time course (TC) experiment under control, cold and heat conditions.  (A) The samples showed strong clustering by genotype (indicated by color of labels) and within each genotype the samples tend to cluster by treatment.  (B) t-SNE plot shows similar clustering based on both genotype and treatment.

\pagebreak
![](plots/sf03.pdf){width=100%}

Figure S3. Expression profiles of cold- and heat-responsive gene clusters.  B73, Mo17 and W22 genes that exhibit significant DE after 1h or 25h stress treatment were used to perform co-expression clustering based on their time-course expression pattern (see methods). The median expression level of control and stress conditions for the genes within each module is shown and the number of B73, Mo17 and W22 genes in each module is indicated. Ranges at each time point represent 25%-75% quantile expression levels.

\pagebreak
![](plots/sf04.pdf){width=80%}

Figure S4. Transcription Factor Family Enrichment in Select DEG Modules. The TFs present in the overall set of up-regulated genes for cold (A) or heat (B) as well as TFs that are present in specific modules were assessed for potential enrichment of specific TF families. Numbers in parentheses indicate total group size, while numbers in tiles indicate how many DEGs of specific TF families were identified in each module.  For modules including at least 3 TF members in a TF family, enrichment was assessed through hypergeometric test (p < 0.05).  Colors in tiles indicate level of enrichment (-log10(p.value)).  Transcription factor family assignments were pulled from Grassius TFDB (Yilmaz et al., 2009).

\pagebreak
![](plots/sf05.pdf){width=80%}

Figure S5. Expression profile of maize Heat Shock Factors (HSF) response to heat stress. Time course expression pattern in B73 was shown for all 29 maize HSFs including:  (A) 5 HSFs exhibit very strong activation at 0.5h of heat stress, but return to relatively normal levels by the 1h time point.  (B) Four HSFs classified as transient activation; (C) One HSF was significantly differentially expressed, but not assigned to a specific cold-upregulated module. (D) Ten HSFs are expressed but not assigned into a cold up-regulation module.  Genes were considered expressed if expression was ≥ 5 cpm in least one treatment x time point.

\pagebreak
![](plots/sf06.pdf){width=100%}

Figure S6. Expression profile of TFs that have previously been reported to play a role in maize response to cold. Out of 235 previously reported cold stress-related TFs, 109 TFs were expressed with ≥ 5 CPM in B73 in at least one sample (treatment x time point), while 43 were differentially expressed in response to cold stress and assigned into cold up-regulated modules including: (A) late activation, (B) early activation, (C) differentially expressed (DE). DE cold up-regulated genes were identified in the hybrid experiment (3 replicates, 1h and 25h after stress) and TF gene response to cold stress in the time-series (1 replicate, 9 time points) may not depict cold up-regulation for each DE gene.


\pagebreak
![](plots/sf07.pdf){width=100%}

Figure S7. Number of enriched motifs found using B73 promoter spaces (“B model”) or B73/Mo17/W22 promoter spaces (“BMW model”) of different sizes and epigenetic filters.  Co-expression modules as well as the entire set of cold-/heat- up-/down-regulated genes were searched for enriched motifs with varying lengths of promoter space and different filters based on methylation or chromatin accessibility.  Identified motifs were compared with known TF binding motifs (cis-BP) and clustered based on sequence similarity.  Numbers in each cell represent the number of non-redundant motifs found using the full set of sequences (raw), using only the unmethylated regions (umr) or using only accessible chromatin regions in leaf (acrL) or ear (acrE) inside different sized promoter spaces. Color indicates the number of motifs identified (the darker, the more motifs found) and corresponds to the number in each cell.  Numbers in parentheses on the left side of plot indicate the number of genes in each co-expression module used for motif mining, while numbers on the right side of plot indicate the total number of non-redundant motifs found for each set of genes after sequence clustering.

\pagebreak
![](plots/sf08.pdf){width=100%}

Figure S8. Enriched motifs include known transcription factor binding sites (TFBSs) as well as novel motifs.  For each set of DEGs up to the top 40 most enriched motifs found using B73 promoter space ("B model") or B73/Mo17/W22 promoter sequences ("BMW model") are shown (p-value for enrichment is indicated by color).  Some sets of DEGs have less than 40 motifs and only the significant ones are shown.  If the enriched motif matched a previously characterized TFBS (Pearson correlation coefficient > 0.8) the name of the transcription factor is shown.  In each case, there are a mixture of previously characterized motifs and novel motifs.  TTS motifs with a "*" sign represent motifs are not present in the top40 TSS motifs identified in the same list of genes.

\pagebreak
![](plots/sf09.pdf){width=80%}

Figure S9. Meta plots of selected stress-responsive motifs in gene promoters.  Two to four motifs from each of the seven co-expression clusters (cold early activation, cold late activation, etc.) with their relative locations in the TSS:-/+2k region cut into 40 bins.  The occurrence of each motif in these bins were recorded and the proportion of genes containing the motif in each bin is plotted.

\pagebreak
![](plots/sf10.pdf){width=60%}

Figure S10. Performance (Area Under ROC Curve, AUROC) of machine learning models predicting cold and heat responsive expression.  Models were trained to predict expression responses to heat or cold using: (A) different sizes of promoter sequences surrounding the TSS; (B) using "binary encoding" (0/1) or "quantitative encoding" (0/1/2/...) of motifs and and (C) using different numbers of mostly enriched motifs as input features. The number of non-redundant motifs used for model training are indicated on top of each set of bars in (C). In each training the average AUROC (N=100 model runs) is show along with the standard deviation.


\pagebreak
![](plots/sf11.pdf){width=100%}

Figure S11.  Relationship between motif enrichment level and feature importance score in different sets of stress responsive genes. Motif enrichment levels are determined by a hypergeometric test using motif occurrences in positive and negative gene sets and ranked from most significant to least significant (x-axis).  Feature importance scores reported by 100 permutations of each random forest model training are shown on the y-axis with error bar indicating 25-75% quantiles. The top 5 feature importance scores are labelled with motif names (known motif) or consensus sequences (novel motif).  All feature importance score estimates are based on the models trained using "+/-2k", "UMR", "top200", "binary" parameters.


\pagebreak
![](plots/sf12.pdf){width=100%}

Figure S12.  Comparison of heat- and cold-response gene expression in B73, Mo17 and W22. (A) At each time point for cold or heat stress we identified DEGs in all three genotypes and show the overlap of DEGs.  The actual expression levels in all three genotypes for the non-redundant set of up-regulated genes is shown in (B).  The genes are classified based on response in B73, Mo17 and W22 (0 indicates not DE and 1 indicates DE in the three genotypes such that ‘111’ indicates DE in B73, Mo17 and W22 while ‘010’ would indicate DE in Mo17 but not B73 or W22.  Note that for some genes classified as having variable responses there are actually similar changes in all genotypes but there are other examples with clear response on some genotypes but not in others.


\pagebreak
![](plots/sf13.pdf){width=100%}

Figure S13. Characterization of genes with variable stress responsive patterns among inbreds.  For each pair of genotypes we identified genes that show significantly stronger (or weaker) response to cold/heat at 1h or 25h time points using a generalized linear model (see methods). (A) Scatterplot showing the log2 fold-change (stress / control at a time point) for compared inbreds (A and B) in x- and y-axis.  Above the diagonal line are genes showing stronger cold response in genotype A while those below the diagonal line represent genes showing stronger genotype B response. Different colors indicate groups of genes with different DE status (e.g., red genes are ones up-regulated in both inbreds (but showing significantly different response), green are genes only up-regulated in genotype B).  (B) The number of genes in each category shown in (A). For each class the response in the two genotypes (A and B) is indicated as up-regulated (“+”), down-regulated (“-”) or not DE (=).  


\pagebreak
![](plots/sf14.pdf){width=100%}

Figure S14. Cis/trans characterization of genes showing different stress response among inbreds. For the subset of genes classified as showing significantly different responses among inbreds that also had SNPs we assessed allele-specific expression in the F1 hybrid.  (A) For each pairwise genotype comparison the proportion of allele 1 change in stress vs control of the F1 (x-axis) was compared to the proportion of the change in expression in the parental genotypes (y-axis).  A maximum likelihood model was applied to classify cis- and trans- inheriance patterns and shown in different colors. (B) The number of genes classified into each type of regulatory pattern in each pairwise comparison.


\pagebreak
![](plots/sf15.pdf){width=70%}

Figure S15. Identification of cis-regulatory variants associated with variable cold responsive pattern in a panel of 25 maize genotypes.  (A) Consistency of log2 fold changes for each gene called between two experiments: the replicated HY experiment with 3 maize inbreds (x-axis) and the NM experiment with 25 maize genotypes (y-axis).  Point density is shown to avoid overplotting.  Linear regression is performed and the adjusted R2 values are marked.  (B) Cis-regulated variable response genes are enriched in having significantly associated local variants.  Shown are proportions of genes with detected local associations for variable response genes under different regulatory patterns.


\pagebreak
![](plots/sf16.pdf){width=90%}

Figure S16. Cis-regulatory variation associated with cold responsive patterns in a respiratory burst oxidase (rboh4, panel A) and a dihydroflavonoid reductase (dfr1, panel B).  Heatmap shows bi-allelic variants (SNP and short indels) within 2kb of the gene with yellow indicating reference (B73) allele and blue indicating alternate allele.  Log2 fold change of each genotype is shown on the right of the heatmap. Below the gene structure plot are locations of the most (top30) enriched cold-responsive motifs (red dots) with the motif name marked, as well as haplotype blocks (dark blue segments) identified using PLINK.  Boxplot on the right shows the top associating variant and the log2fc distributions of genotypes carrying the two alleles.


\pagebreak

```{r, message=FALSE, echo=FALSE}
dirf = '/home/springer/zhoux379/projects/stress/data/95_figures/plots'
options(tinytex.verbose = TRUE)
```

Table S1. Sample statistics.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirf, 'st1.rds'))
```

