---
title: "maize stress manuscript figures and tables"
output:
  pdf_document:
    citation_package: natbib
    keep_tex: false
    fig_caption: true
    latex_engine: lualatex
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
date: "`r format(Sys.time(), '%B %d, %Y')`"
fontsize: 11pt
geometry: "left=0.5in,right=0.5in,top=0.8in,bottom=0.5in"
# spacing: double
# biblio-style: apsr
header-includes:
- \PassOptionsToPackage{table}{xcolor}
- \usepackage{fontspec}
- \setmainfont[Scale=MatchUppercase]{TeX Gyre Heros}
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
---
\pagenumbering{gobble}

```{r, message=FALSE, echo=FALSE}
dirf = '/home/springer/zhoux379/projects/stress/data/95_figures/plots'
```

\hfill

![]($st/data/01_exp_design/design.pdf){width=100%}

Figure S1. Experimental design used for generation of RNA-Seq data.  Three different experiments were conducted to assess the changes in gene expression in response to cold or heat stress in 14 day maize seedlings.  In experiment 1 (time course - TC) one replicate for three inbreds was collected at each of the indicated time-points.  The panel on the right shows the temperature readings from a sensor that was with the plants throughout the experiment and shows the time of light / dark (gray shading indicates time without lights).  In experiment 2 (inbred - hybrid - HY, second row) three biological replicates were sampled from three maize inbreds and their F1 hybrids at time 0 and two time points during the stress. The final row shows experiment 3 in which a single biological replicate was sampled for a panel of 25 genotypes under control and cold conditions at two time points.


\pagebreak
![](plots/f1.pdf){width=100%}

Figure 1. Identification of genes that are differentially expressed in response to heat or cold stress. (A) t-SNE clustering of samples from the HY experiment under control, cold and heat conditions.  The genotypes are indicated by different symbols/colors and the conditions for each set of samples are indicated in the plot. (B) Number of differentially expressed genes (DEGs) under cold and heat conditions at 1 and 25 hour time points. For each time point the number of genes that showed differential expression (see Methods) relative to the control sampled at time 0 (the onset of the stress; time0_control) is shown as well as the number of genes that showed differential expression relative to the control sample collected at matching time point (i.e., 1h or 25h; timeM_control). Numbers inside red bar represent genes that are differentially expressed in both comparisons.

\pagebreak
![](plots/sf02.pdf){width=100%}

Figure S2. Hierarchical (A) and t-SNE (B) clustering of all samples from the time course (TC) experiment under control, cold and heat conditions.  (A) The samples showed strong clustering by genotype (indicated by color of labels) and within each genotype the samples tend to cluster by treatment.  (B) t-SNE plot shows similar clustering based on both genotype and treatment.

\pagebreak
![](plots/f2.pdf){width=100%}

Figure 2. Expression profile and GO enrichment of cold- and heat-responsive gene clusters.  B73, Mo17 and W22 genes that exhibit significant DE after 1 hour or 25 hours stress treatment were used to perform co-expression clustering based on their time-course expression pattern (see methods).  (A) Expression pattern for the most abundant co-expression clusters that represent early, late or transient changes upon cold or heat treatment.  The median expression level of control and stress conditions for the genes within each module is shown and the number of B73, Mo17 and W22 genes in each module is indicated. Ranges at each time point represent 25%-75% quantile expression levels. (B) For the set of all up-regulated genes under cold or heat conditions we determined the number of TFs and assessed GO enrichment.  Two highly enriched GO terms (with significance p < 0.01) are shown for each module, and the significance level is indicated.

\pagebreak
![](plots/sf03.pdf){width=100%}

Figure S3. Expression profiles of cold- and heat-responsive gene clusters.  B73, Mo17 and W22 genes that exhibit significant DE after 1 hour or 25 hours stress treatment were used to perform co-expression clustering based on their time-course expression pattern (see methods). The median expression level of control and stress conditions for the genes within each module is shown and the number of B73, Mo17 and W22 genes in each module is indicated. Ranges at each time point represent 25%-75% quantile expression levels.

\pagebreak
![](plots/sf04.pdf){width=80%}

Fig S4. Transcription Factor Family Enrichment in Select DEG Modules. The TFs present in the overall set of up-regulated genes for cold or heat as well as TFs that are present in specific modules were assessed for potential enrichment of specific TF families. Numbers in parentheses indicate total group size, while numbers in tiles indicate how many DEGs of specific TF families were identified in each module.  For modules including at least 3 TF members in a TF family, enrichment was assessed through hypergeometric test (p < 0.05). Transcription factor family assignments were pulled from Grassius TFDB (Yilmaz et al., 2009).

\pagebreak
![](plots/sf05.pdf){width=80%}

Figure S5. Characterization of maize Heat Shock Factors (HSF) response to heat stress. Time course expression pattern was shown for all 29 maize HSFs including:  (A) 5 HSFs exhibit very strong activation at 0.5 hours of heat stress, but return to relatively normal levels by the 1 hour time point.  (B) 4 HSFs classified as transient activation; (C) 1 HSF was significantly differentially expressed, but not assigned to a specific cold-upregulated module. (D) 10 HSFs are expressed but not assigned into a cold up-regulation module.  Genes were considered expressed if expression was ≥ 5 cpm in least one treatment x time point.

\pagebreak
![](plots/sf06.pdf){width=100%}

Figure S6. Characterization of TFs that have previously been reported to play a role in maize response to cold. Out of 235 previously reported cold stress-related TFs, 109 TFs were expressed with ≥ 5 CPM for at least one sample (treatment x time point), while 43 were differentially expressed in response to cold stress and assigned into cold up-regulated modules including: (A) late activation, (B) early activation, (C) differentially expressed (DE). DE cold up-regulated genes were identified in the hybrid experiment (3 replicates, 1h and 25h after stress) and TF gene response to cold stress in the time-series (1 replicate, 9 time points) may not depict cold up-regulation for each DE gene.

\pagebreak
![](plots/f3.pdf){width=60%}

Figure 3. Identification of enriched motifs in cold- and heat-responsive genes. (A) Varying potential ‘promoter’ sequence spaces were utilized to search for motifs that are enriched in different sets of genes.  The schematic indicates a representative gene with the transcription start site (TSS) indicated.  The potential regions include different lengths of sequences upstream the promoter (-500, -1k, -2k) as well as sets of sequence that include both upstream and downstream sequence (i.e., +/-500).  In addition, for each of these potential regions we also subset the sequence to only include regions that are unmethylated (UMR - unmethylated regions) or that are classified as accessible based on ATAC analysis (ACR - accessible chromatin region) in leaf or ear tissue (Ricci et al., 2019).  (B) Co-expression modules as well as the entire set of cold-/heat- up-/down-regulated genes were used to search for enriched motifs (see Methods).  The number of non-redundant motifs found using +/- 2k promoter sequence as well as different epigenetic filters (i.e., UMRs, ACRs) were shown. The numbers on the right side of the plot indicate the total number of non-redundant motifs found for each set of genes using +/-2k promoter space as well as the other promoter spaces of different sizes (shown in Figure S7).

\pagebreak
![](plots/sf07.pdf){width=100%}

Figure S7. Number of enriched motifs found using promoter spaces of different size and epigenetic filters.  Co-expression modules as well as the entire set of cold-/heat- up-/down-regulated genes were searched for enriched motifs with varying lengths of promoter space and different filters based on methylation or chromatin accessibility.  Identified motifs were compared with known TF binding motifs (cis-BP) and clustered based on sequence similarity.  The number of non-redundant motifs found using the full set of sequences (raw) in the specific region is shown as well as the number of motifs when only the unmethylated (umr) or accessible chromatin (acr) in this region was utilized. The numbers on the right side of the plot indicate the total number of non-redundant motifs found for each set of genes after sequence clustering.

\pagebreak
![](plots/sf08.pdf){width=100%}

Figure S8. Enriched motifs include known transcription factor binding sites (TFBSs) as well as novel motifs.  For each set of DEGs up to the top 40 most enriched motifs are shown (p-value for enrichment is indicated by color).  Some sets of DEGs have less than 40 motifs and only the significant ones are shown.  If the enriched motif matched a previously characterized TFBS (Pearson correlation coefficient > 0.8) the name of the transcription factor is shown.  In each cases there are a mixture of previously characterized motifs and novel motifs.  TTS motifs with a "*" sign represent motifs not present in the top40 TSS motifs identified in the same list of genes.

\pagebreak
![](plots/f4.pdf){width=60%}

Figure 4. Performance (Area Under ROC Curve, AUROC) of machine learning models predicting cold and heat responsive expression.  Models were trained to predict expression responses to heat or cold using: (A) different windows near the TSS and TTS; (B) with all sequence (red) or UMR regions only (blue); (C) using "binary encoding" (0/1) or "quantitative encoding" (0/1/2/...) of motifs and (D) using different numbers of mostly enriched motifs as input features. In each training the average AUROC (N=100 model runs) is show along with the standard deviation.


\pagebreak
![](plots/sf09.pdf){width=60%}

Figure S9. Prediction of cold and heat responsive expression.  (A) Machine learning models were trained to predict expression responses to heat or cold.  The average F1 score (N=100 model runs) is shown for models trained using all cold or heat up- or down-regulated genes.  Models were run using different windows near the TSS and with all sequence (red) or UMR regions only (blue).  The standard deviation for model performance is shown for each category.  (B) Models were trained for specific up-regulated co-expression modules with at least 20 enriched motifs and compared to the performance of models trained using all up-regulated genes.

\pagebreak
![](plots/f5.pdf){width=100%}

Figure 5. Characterization of genes with variable stress responsive patterns among inbreds.  (A) A Venn diagram is used to assess genes that are up-regulated in response to 25h of cold stress for B73, Mo17 and W22.  The overlap of DE genes at other time points is shown in Figure S11A.  (B) For genes that show significantly stronger (or weaker) response to cold at 25h in B73 compared to Mo17 we show the log2 fold-change (cold 25h / control 25h) for both inbreds. The classification of differential responses for other genotype contrasts and timepoints is provided in Figure S12.  Above the diagonal line are genes showing stronger cold response in Mo17 while those below the diagonal line represent genes showing stronger B73 response. Different colors indicate groups of genes with different DE status (e.g., red genes are ones up-regulated in both inbreds (but showing significantly different response), green are genes only up-regulated in Mo17).   The number of genes in each category is shown in (C). For each class the response in the two genotypes (A and B) is indicated as up-regulated (“+”), down-regulated (“-”) or not DE (=).  (D) For the subset of genes classified as having a response in only one of the two genotypes that also had SNPs we assessed allele-specific expression in the F1 hybrid.  The proportion of allele 1 (B73) change in stress vs control of the F1 (x-axis) was compared to the proportion of the change in expression in the parental genotypes (y-axis).  A maximum likelihood model was applied to classify cis- and trans- inheriance patterns and shown in different colors. The number of genes classified into each type of regulatory pattern for response to abiotic stress are shown in (E).  Similar analyses for other genotypes, stress and time points are shown in Figure S13.

\pagebreak

Figure S10. TF meta plot

\pagebreak
![](plots/sf11.pdf){width=100%}

Figure S11.  Comparison of heat- and cold-response gene expression in B73, Mo17 and W22. (A) At each time point for cold or heat stress we identified DEGs in all three genotypes and show the overlap of DEGs.  The actual expression levels in all three genotypes for the non-redundant set of genes in shown in (B).  The genes are classified based on response in B73, Mo17 and W22 (0 indicates not DE and 1 indicates DE in the three genotypes such that ‘111’ indicates DE in B73, Mo17 and W22 while ‘010’ would indicate DE in Mo17 but not B73 or W22.  Note that for some genes classified as having variable responses there are actually similar changes in all genotypes but there are other examples with clear response on some genotypes but not in others.


\pagebreak
![](plots/sf12.pdf){width=100%}

Figure S12. Characterization of genes with variable stress responsive patterns among inbreds.  For each pair of genotypes we identified genes that show significantly stronger (or weaker) response to cold/heat at 1h or 25h time points using a generalized linear model (see methods). (A) Scatterplot showing the log2 fold-change (stress / control at a time point) for compared inbreds (A and B) in x- and y-axis.  Above the diagonal line are genes showing stronger cold response in genotype A while those below the diagonal line represent genes showing stronger genotype B response. Different colors indicate groups of genes with different DE status (e.g., red genes are ones up-regulated in both inbreds (but showing significantly different response), green are genes only up-regulated in genotype B).  (B) The number of genes in each category shown in (A). For each class the response in the two genotypes (A and B) is indicated as up-regulated (“+”), down-regulated (“-”) or not DE (=).  


\pagebreak
![](plots/sf13.pdf){width=100%}

Figure S13. Cis/trans characterization of genes showing different stress response among inbreds. For the subset of genes classified as showing significantly different response among inbreds that also had SNPs we assessed allele-specific expression in the F1 hybrid.  (A) For each pariwise genotype comparison the proportion of allele 1 change in stress vs control of the F1 (x-axis) was compared to the proportion of the change in expression in the parental genotypes (y-axis).  A maximum likelihood model was applied to classify cis- and trans- inheriance patterns and shown in different colors. (B) The number of genes classified into each type of regulatory pattern in each pariwise comparison.


\pagebreak




