---
title: "maize stress manuscript figures and tables"
output:
  pdf_document:
    citation_package: natbib
    keep_tex: false
    fig_caption: true
    latex_engine: lualatex
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
date: "`r format(Sys.time(), '%B %d, %Y')`"
fontsize: 11pt
geometry: "left=0.5in,right=0.5in,top=0.8in,bottom=0.5in"
# spacing: double
# biblio-style: apsr
header-includes:
- \PassOptionsToPackage{table}{xcolor}
- \usepackage{fontspec}
- \setmainfont[Scale=MatchUppercase]{TeX Gyre Heros}
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
---
\pagenumbering{gobble}

```{r, message=FALSE, echo=FALSE}
dirf = '/home/springer/zhoux379/projects/stress/data/95_figures/plots'
```

\pagebreak
![](plots/f1.pdf){width=100%}

Figure 1. Identification of genes that are differentially expressed in response to heat or cold stress. (A) t-SNE clustering of samples from the hybrid (HY) experiment under control, cold and heat conditions.  The genotypes are indicated by different symbols/colors and the conditions for each set of samples are indicated in the plot. (B) The number of differentially expressed genes (DEGs) under cold and heat conditions at 1 and 25 hour time points. For each time point the number of genes that showed differential expression (see Methods) relative to the control sampled at time 0 (the onset of the stress; time0_control) is shown as well as the number of genes that showed differential expression relative to the control sample collected at matching time point (i.e., 1h or 25h; timeM_control). Numbers inside red bars represent genes that are differentially expressed in both comparisons.

\pagebreak
![](plots/f2.pdf){width=100%}

Figure 2. Expression profiles and GO enrichment of cold- and heat-responsive gene clusters.  B73, Mo17 and W22 genes that exhibit significant DE after 1 hour or 25 hours stress treatment were used to perform co-expression clustering based on their time-course expression pattern (see methods).  (A) The expression pattern for the most abundant co-expression clusters that represent early, late or transient changes upon cold or heat treatment.  The median expression level of control and stress conditions for the genes within each module is shown and the number of B73, Mo17 and W22 genes in each module is indicated. Ranges at each time point represent 25%-75% quantile expression levels. (B) For the set of all up-regulated genes under cold or heat conditions we determined the number of TFs and assessed GO enrichment.  Two highly enriched GO terms (with significance p < 0.01) are shown for each module, and the significance level is indicated.

\pagebreak
![](plots/f3.pdf){width=70%}

Figure 3. Identification of enriched motifs in cold- and heat-responsive genes. (A) Varying potential ‘promoter’ sequence spaces were utilized to search for motifs that are enriched in different sets of genes.  The schematic indicates a representative gene with the transcription start site (TSS) indicated.  The potential regions include different lengths of sequences upstream the promoter (-500, -1k, -2k) as well as sets of sequence that include both upstream and downstream sequence (i.e., +/-500).  In addition, for each of these potential regions we also subset the sequence to only include regions that are unmethylated (UMR - unmethylated regions) or that are classified as accessible based on ATAC-seq analysis (ACR - accessible chromatin region) in leaf or ear tissue (Ricci et al., 2019).  (B) The number of non-redundant motifs found using B73 promoter spaces (“B model”) or B73/Mo17/W22 promoter spaces (“BMW model”) and different search spaces is indicated. Colors indicate numbers of motifs identified (the darker, the more motifs found) and correspond to numbers in each cell.  Numbers in parentheses on the left side of the heatmap indicate the number of genes used for motif mining; while numbers on the right side of the plot indicate the total number of non-redundant motifs found for each set of genes.  More detailed search results using different promoter sizes along with different chromatin properties (UMRs, ACRs) are shown in Figure S7.

\pagebreak
![](plots/f4.pdf){width=60%}

Figure 4. Performance (Area Under ROC Curve, AUROC) comparison of different machine learning models predicting cold and heat responsive expression.  (A) The performance comparison of models trained using small promoter spaces (-500, +500, -/+500) against larger promoter spaces (-2k, +2k, -/+2k).  In each case the same chromatin filters (UMR) and number of features (top100) were used for training. (B) The performance comparison of models trained using the full set of DEGs against models trained using specific co-expression clusters.  In each case same sized promoter spaces (-/+2k), chromatin filters (UMR) and number of features (top100) were used. (C) The performance comparison of models using all genomic sequence ("all"), UMR regions only ("umr") or using leaf accessible regions only ("acrL") with the same sized promoter spaces (-/+2k) and number of features (top100).  In each training average AUROC (N=100 model runs) is shown along with the standard deviation.  Pairwise comparisons were made using t-test with significance levels indicated (\*: 0.05; \*\*: 0.01; \*\*\*: 0.001)

\pagebreak
![](plots/f5.pdf){width=100%}

Figure 5. Characterization of genes with variable stress responsive patterns among inbreds.  (A) A Venn diagram is used to compare the genes that are up-regulated in response to 25h of cold stress for B73, Mo17 and W22.  The overlap of DE genes at other time points is shown in Figure S11A.  (B) For genes that show significantly stronger (or weaker) response to cold at 25h in B73 compared to Mo17 we show the log2 fold-change (cold 25h / control 25h) for both inbreds. The classification of differential responses for other genotype contrasts and timepoints is provided in Figure S12. Post-hoc tests were used to classify (colors) genes with varying differential expression between genotypes; red indicates up-regulation in both inbreds (with different response levels), green indicates genes only up-regulated in Mo17, etc.   The number of genes in each category is shown in (C). For each class the response in the two genotypes (A and B) is indicated as up-regulated (“+”), down-regulated (“-”) or not DE (=).  (D) For the subset of genes classified as having a response in only one of the two genotypes that also had SNPs we assessed allele-specific expression in the F1 hybrid.  The proportion of allele 1 (B73) change in stress vs control of the F1 (x-axis) was compared to the proportion of the change in expression in the parental genotypes (y-axis).  A maximum likelihood model was applied to classify cis- and trans- inheritance patterns and these classifications are shown in different colors. The number of genes classified into each type of regulatory pattern for response to abiotic stress are shown in (E).  Similar analyses for other genotypes, stress and time points are shown in Figure S13.

\pagebreak
![](plots/f6.pdf){width=100%}

Figure 6. Cross-genotype performance of machine learning models predicting cold or heat responsive expression.  Models were trained only using B73 sequence and DE labels ("B model") or data from all three genotypes ("BMW model"). (A) AUROC for models predicting stress responsive expression in B73, Mo17 and W22.  Average AUROC (N=100 model permutations) is shown along with the standard deviation for both B and BMW models. (B) Model prediction accuracy for genes showing consistent ("Both respond") or variable ("Only B73 or Mo17/W22 respond") response patterns among genotypes.  In each observed category the number and proportion of predictions were marked in the plot with the correct predictions highlighted with red boxes.  (C) Dissection of regulatory patterns for genes showing variable response patterns among genotypes.  Variable response genes were first grouped by whether model prediction agrees with observed status ("Correct" if the model correctly predicts one genotype responds but the other does not, "Incorrect" if the model predicts oppositely, "Both respond" and "Neither respond" if the model predicts both or neither genotypes respond - although in reality only one genotype responds). Then within each group the number of proportion of different regulatory patterns ("cis", "trans", etc) were marked.

\pagebreak
![](plots/f7.pdf){width=80%}

Figure 7. Identification of cis-regulatory variants associated with variable cold responsive pattern in a panel of 25 maize genotypes. (A) The proportion of ~2000 variable response genes showing uni-modal, bi-modal and multi-modal distributions of the log2 fold change ratios (cold/control);  (B) Examples of two genes showing uni-modal log2fc ratios and two showing bi-modal ratio distributions;  (C) Cis-regulatory variation associated with cold responsiveness in a maize mitochondrial transcription termination factor (mTERF17).  Heatmap shows biallelic variants (SNP and short indels) within 2kb of the gene with yellow indicating reference (B73) allele and blue indicating alternate allele.  Log2 fold change of each genotype is shown on the right of the heatmap. Below the gene structure plot are locations of the most (top30) enriched cold-responsive motifs (red dots ) with the motif name marked, as well as haplotype blocks (dark blue segments) identified using PLINK.  The boxplot on the right shows the top associating variant and the log2fc distributions of genotypes carrying the two alleles.


\pagebreak

```{r, message=FALSE, echo=FALSE}
dirf = '/home/springer/zhoux379/projects/stress/data/95_figures/plots'
options(tinytex.verbose = TRUE)
```
